Code/GraphMol/MolStandardize/Pipeline.cpp

```diff
@@ -0,0 +1,593 @@
+//
+//  Copyright (C) 2023 Novartis Biomedical Research
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <cmath>
+#include <regex>
+#include <sstream>
+#include "Pipeline.h"
+#include "Validate.h"
+#include "Metal.h"
+#include "Normalize.h"
+#include "Charge.h"
+#include "Fragment.h"
+#include <RDGeneral/FileParseException.h>
+#include <GraphMol/FileParsers/FileParsers.h>
+#include <GraphMol/SmilesParse/SmilesParse.h>
+#include <GraphMol/SmilesParse/SmilesWrite.h>
+#include <GraphMol/Chirality.h>
+
+namespace RDKit {
+namespace MolStandardize {
+
+void PipelineResult::append(PipelineStatus newStatus, const std::string &info) {
+  status = static_cast<PipelineStatus>(status | newStatus);
+  log.push_back({newStatus, info});
+}
+
+PipelineResult Pipeline::run(const std::string &molblock) const {
+  PipelineResult result;
+  result.status = NO_EVENT;
+  result.inputMolBlock = molblock;
+
+  // parse the molblock into an RWMol instance
+  result.stage = PARSING_INPUT;
+  RWMOL_SPTR mol = parse(molblock, result);
+  if (!mol || ((result.status & PIPELINE_ERROR) != NO_EVENT &&
+               !options.reportAllFailures)) {
+    return result;
+  }
+
+  RWMOL_SPTR_PAIR output;
+
+  if (mol->getNumAtoms() == 0 && options.allowEmptyMolecules) {
+    output = {mol, mol};
```

Comment:
I didn't think it was necessary because `RDKitValidation` reports an error if the molecule has no atoms and the conditional logic here was mainly motivated by preventing that.

I could maybe cleanup the execution flow a bit here, if `RDKitValidation` was modified to support an `allowEmptyMolecules` flag, or if the test on the number of atoms was removed from it, and replaced in the pipeline code by the conditional execution of `NoAtomValidation`. What do you think?

Reply:
The standardization pipeline could also be used with a different validation work flow.
I think that if you are going to be checking `allowEmptyMolecules` at all here, then you need to handle the situation where it's false as well.

Reply:
I can move this check into the validation workflow if we make `RDKitValidation` configurable. Otherwise, if we assume that the work flow can be different and therefore more directly responsible for allowing or rejecting empty molecules, I am tempted to just remove `allowEmptyMolecules` from the pipeline options. I don't actually need it, and after this discussion I'm thinking it wasn't a great idea (a kind of yagni).

If you think it may be useful to have, I would change `RDKitValidation` to optionally allow empty molecules, otherwise I'd just remove `allowEmptyMolecules` and the code that depends on it.